pyquil/quil.py

```diff
@@ -981,3 +983,36 @@ def percolate_declares(program: Program) -> Program:
     p._defined_gates = program._defined_gates
 
     return p
+
+
+def validate_protoquil(program: Program) -> None:
```

Comment:
@karalekas recently pointed out that there's one extra thing to check: gate parameters involving classical expressions should always look like

    d + c_1 var_1 + c_2 var_2 + ... + c_k var_k,

where each c_j is a power of 2. It's a little obnoxious to check this, since doing a good job involves flattening an arithmetic expression to be in that form, but it is literally part of the current definition of ProtoQuil.

Reply:
presumably there is some code further down in the stack we can ~~steal~~ use that does this verification prior to execution

Reply:
There sure is; I wrote it. I can PM you both with where it lives.

Reply:
This will be a follow up PR